File: python/walayer/webhook.py
"""Webhook signature verification.

Matches the server scheme (docs/04 §8.3): ``X-Signature: v1,sha256=<hex>`` where
hex = HMAC-SHA256(secret, "{timestamp}.{rawBody}"). Verify against the RAW body —
re-serialized JSON breaks the HMAC.
"""
from __future__ import annotations

import hashlib
import hmac
import time
from dataclasses import dataclass
from typing import Optional, Union

SIGNATURE_SCHEME = "v1"
DEFAULT_TOLERANCE_SECONDS = 300


@dataclass(frozen=True)
class VerifyResult:
    valid: bool
    reason: Optional[str] = None  # "format" | "timestamp" | "signature"


def verify_webhook(
    raw_body: str,
    signature: Optional[str],
    timestamp: Union[int, str, None],
    secret: str,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
    now_seconds: Optional[int] = None,
) -> VerifyResult:
    """Constant-time verification with a replay window."""
    prefix = f"{SIGNATURE_SCHEME},sha256="
    if not signature or not signature.startswith(prefix):
        return VerifyResult(False, "format")

    try:
        ts = int(timestamp) if timestamp is not None else None
    except (TypeError, ValueError):
        ts = None
    now = now_seconds if now_seconds is not None else int(time.time())
    if ts is None or abs(now - ts) > tolerance_seconds:
        return VerifyResult(False, "timestamp")

    mac = hmac.new(secret.encode("utf-8"), f"{ts}.{raw_body}".encode("utf-8"), hashlib.sha256).hexdigest()
    expected = f"{prefix}{mac}"
    if not hmac.compare_digest(signature, expected):
        return VerifyResult(False, "signature")
    return VerifyResult(True)

File: php/src/Webhook.php
<?php

declare(strict_types=1);

namespace WALayer;

/**
 * Webhook signature verification.
 *
 * Matches the server scheme (docs/04-api-spec.md §8.3): the request carries
 * `X-Signature: v1,sha256=<hex>` and `X-Timestamp: <unix seconds>`, where
 * hex = HMAC-SHA256(secret, "{timestamp}.{rawBody}").
 *
 * Verify against the RAW body — re-serialising the JSON breaks the HMAC.
 */
final class Webhook
{
    public const SIGNATURE_SCHEME = 'v1';
    public const DEFAULT_TOLERANCE_SECONDS = 300;

    private const PREFIX = self::SIGNATURE_SCHEME . ',sha256=';

    /**
     * Constant-time verification with a replay window enforced in both
     * directions: a timestamp far in the past is a replay, one far in the
     * future is a forged or badly-skewed sender. Neither is accepted.
     *
     * @param string          $rawBody          the request body exactly as received
     * @param string|null     $signature        the `X-Signature` header
     * @param int|string|null $timestamp        the `X-Timestamp` header (unix seconds)
     * @param string          $secret           the endpoint's signing secret
     * @param int             $toleranceSeconds half-width of the accepted window
     * @param int|null        $nowSeconds       override for tests; defaults to time()
     */
    public static function verify(
        string $rawBody,
        ?string $signature,
        int|string|null $timestamp,
        string $secret,
        int $toleranceSeconds = self::DEFAULT_TOLERANCE_SECONDS,
        ?int $nowSeconds = null
    ): VerifyResult {
        if ($signature === null || !\str_starts_with($signature, self::PREFIX)) {
            return VerifyResult::fail(VerifyResult::REASON_FORMAT);
        }

        if (!\is_int($timestamp) && (!\is_string($timestamp) || !\preg_match('/^-?\d+$/', $timestamp))) {
            return VerifyResult::fail(VerifyResult::REASON_TIMESTAMP);
        }
        $ts = (int) $timestamp;
        $now = $nowSeconds ?? \time();
        // Both directions: stale replays AND timestamps from the future.
        if (\abs($now - $ts) > $toleranceSeconds) {
            return VerifyResult::fail(VerifyResult::REASON_TIMESTAMP);
        }

        $expected = self::PREFIX . \hash_hmac('sha256', $ts . '.' . $rawBody, $secret);
        if (!\hash_equals($expected, $signature)) {
            return VerifyResult::fail(VerifyResult::REASON_SIGNATURE);
        }

        return VerifyResult::ok();
    }

    /**
     * Sign a payload the way the server does. Present so tests and local
     * webhook replays can produce a real header; production code only ever
     * needs {@see Webhook::verify()}.
     */
    public static function sign(string $rawBody, int $timestamp, string $secret): string
    {
        return self::PREFIX . \hash_hmac('sha256', $timestamp . '.' . $rawBody, $secret);
    }
}
